core/deployment/src/main/java/io/quarkus/deployment/util/AsmUtil.java

```diff
@@ -68,49 +72,168 @@
      * public class Foo<T> {
-     *  public <R> List<R> method(int a, T t){...} 
+     *  public <R> List<R> method(int a, T t){...}
      * }
      * }
      * </pre>
-     * 
+     *
      * This will return <tt>&lt;R:Ljava/lang/Object;>(ILjava/lang/Integer;)Ljava/util/List&lt;TR;>;</tt> if
      * your {@code typeArgMapper} contains {@code T=Ljava/lang/Integer;}.
-     * 
+     *
      * @param method the method you want the signature for.
      * @param typeArgMapper a mapping between type argument names and their bytecode signature.
      * @return a bytecode signature for that method.
      */
     public static String getSignature(MethodInfo method, Function<String, String> typeArgMapper) {
-        List<Type> parameters = method.parameters();
+        // for grammar, see JVMS 16, chapter 4.7.9.1. Signatures
 
-        StringBuilder signature = new StringBuilder("");
-        for (TypeVariable typeVariable : method.typeParameters()) {
-            if (signature.length() == 0)
-                signature.append("<");
-            else
-                signature.append(",");
-            signature.append(typeVariable.identifier()).append(":");
-            // FIXME: only use the first bound
-            toSignature(signature, typeVariable.bounds().get(0), typeArgMapper, false);
+        StringBuilder signature = new StringBuilder();
+
+        if (!method.typeParameters().isEmpty()) {
+            signature.append('<');
+            for (TypeVariable typeParameter : method.typeParameters()) {
+                typeParameter(typeParameter, signature, typeArgMapper);
+            }
+            signature.append('>');
         }
-        if (signature.length() > 0)
-            signature.append(">");
-        signature.append("(");
-        for (Type type : parameters) {
+
+        signature.append('(');
+        for (Type type : method.parameters()) {
             toSignature(signature, type, typeArgMapper, false);
         }
-        signature.append(")");
+        signature.append(')');
+
         toSignature(signature, method.returnType(), typeArgMapper, false);
+
+        if (hasThrowsSignature(method)) {
+            for (Type exception : method.exceptions()) {
+                signature.append('^');
+                toSignature(signature, exception, typeArgMapper, false);
+            }
+        }
+
         return signature.toString();
     }
 
+    private static void typeParameter(TypeVariable typeParameter, StringBuilder result,
+            Function<String, String> typeArgMapper) {
+        result.append(typeParameter.identifier());
+
+        if (hasImplicitObjectBound(typeParameter)) {
+            result.append(':');
+        }
+        for (Type bound : typeParameter.bounds()) {
+            result.append(':');
+            toSignature(result, bound, typeArgMapper, false);
+        }
+    }
+
+    private static boolean hasImplicitObjectBound(TypeVariable typeParameter) {
+        // TODO is there a better way? :-/
```

Comment:
@n1hility any help here? This is the only thing I wasn't able to reconstruct from Jandex's public API.